src/Analyser/MutatingScope.php

```diff
@@ -2008,6 +2008,177 @@ private function resolveType(Expr $node): Type
 			if ($this->reflectionProvider->hasConstant($node->name, $this)) {
 				/** @var string $resolvedConstantName */
 				$resolvedConstantName = $this->reflectionProvider->resolveConstantName($node->name, $this);
+				// core, https://www.php.net/manual/en/reserved.constants.php
+				if ($resolvedConstantName === 'PHP_VERSION') {
+					return new IntersectionType([
+						new StringType(),
+						new AccessoryNonEmptyStringType(),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_MAJOR_VERSION') {
+					return IntegerRangeType::fromInterval(5, null);
+				}
+				if ($resolvedConstantName === 'PHP_MINOR_VERSION') {
+					return IntegerRangeType::fromInterval(0, null);
+				}
+				if ($resolvedConstantName === 'PHP_RELEASE_VERSION') {
+					return IntegerRangeType::fromInterval(0, null);
+				}
+				if ($resolvedConstantName === 'PHP_VERSION_ID') {
+					return IntegerRangeType::fromInterval(50207, null);
+				}
+				if ($resolvedConstantName === 'PHP_ZTS') {
+					return new UnionType([
+						new ConstantIntegerType(0),
+						new ConstantIntegerType(1),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_DEBUG') {
+					return new UnionType([
+						new ConstantIntegerType(0),
+						new ConstantIntegerType(1),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_MAXPATHLEN') {
+					return IntegerRangeType::fromInterval(1, null);
+				}
+				if ($resolvedConstantName === 'PHP_OS') {
+					return new IntersectionType([
+						new StringType(),
+						new AccessoryNonEmptyStringType(),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_OS_FAMILY') {
+					return new UnionType([
+						new ConstantStringType('Windows'),
+						new ConstantStringType('BSD'),
+						new ConstantStringType('Darwin'),
+						new ConstantStringType('Solaris'),
+						new ConstantStringType('Linux'),
+						new ConstantStringType('Unknown'),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_SAPI') {
+					return new UnionType([
+						new ConstantStringType('apache'),
+						new ConstantStringType('apache2handler'),
+						new ConstantStringType('cgi'),
+						new ConstantStringType('cli'),
+						new ConstantStringType('cli-server'),
+						new ConstantStringType('embed'),
+						new ConstantStringType('fpm-fcgi'),
+						new ConstantStringType('litespeed'),
+						new ConstantStringType('phpdbg'),
+						new IntersectionType([
+							new StringType(),
+							new AccessoryNonEmptyStringType(),
+						]),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_EOL') {
+					return new UnionType([
+						new ConstantStringType("\n"),
+						new ConstantStringType("\r\n"),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return IntegerRangeType::fromInterval(1, null);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MIN') {
+					return IntegerRangeType::fromInterval(null, -1);
+				}
```

Comment:
or alternatively we just fix this constant to 2 values, the actual value PHP_INT_MAX has on 32 bit php and 64 bit php

Reply:
Yes, this should be those two values for 32bit/64bit.

Reply:
a union type of those? I think the initial problem was that the 64 bit value is not settable on 32 bit as it would be an overflow already. or am I misunderstanding this?

Reply:
Yeah, so maybe a single value on 32bit and union on 64bit.

Reply:
I wish there were stats on the share of users still running php on 32 bits... Couldn't find them if they exists

Reply:
Afaik 32 bit is commonly used on raspberry PI

Reply:
... I actually want to see someone trying to run Psalm/PHPStan on a big codebase from a raspberry PI now! (sorry for the Off Topic)

Reply:
Currently adding that 32/64 bit switch for the min/max constants  - there is no 32 bit CI server, right? :D should I create 32/64 bit test files for those 2 constants or just go with the 64 bit "default" case?

Reply:
Code like nextcloud/owncloud etc. is often run on raspi and therefore is coded 32bit compatible. 

I guess we need the unnion of both values here to allow projects to write special cases for 32bit. 
Its not about running phpstan on 32bit

Reply:
Agreed, don't get me wrong, rasing this and the union type was a good idea! I was also trying to find out if it currently runs on 32 bit. It showed that it runs just fine on 32 bit, but tests did not, which I fixed in #957. So now, it is fully compatible there and I won't/can't break it any more :)

Sooo, I adapted this PR, as suggested by Ondrej, to return the union type on 64 bit and the single, smaller value, on 32 bit, to not cause a fatal error there. To keep the tests running I simply added a few conditionals where needed for now.
Maybe that's fine, let me know what you all think.

TL/DR: adapted, check the 2 int size specific test files :)